production/ksonnet/loki/multi-zone.libsonnet

```diff
@@ -0,0 +1,201 @@
+  ingester_zone_b_args:: {},
+  ingester_zone_c_args:: {},
+
+  newIngesterZoneContainer(zone, zone_args)::
+    local zone_name = 'zone-%s' % zone;
+
+    $.ingester_container +
+    container.withArgs($.util.mapToFlags(
+      $.ingester_args + zone_args
+    )) +
+    container.withEnvMixin([{ name: 'AVAILABILITY_ZONE', value: zone_name }]),
+
+  newIngesterZoneStatefulSet(zone, container)::
+    local name = 'ingester-zone-%s' % zone;
+
+    $.newIngesterStatefulSet(name, container) +
+    statefulSet.mixin.metadata.withLabels({ 'rollout-group': 'ingester' }) +
+    statefulSet.mixin.metadata.withAnnotations({ 'rollout-max-unavailable': std.toString($._config.multi_zone_ingester_max_unavailable) }) +
+    statefulSet.mixin.spec.template.metadata.withLabels({ name: name, 'rollout-group': 'ingester' }) +
+    statefulSet.mixin.spec.selector.withMatchLabels({ name: name, 'rollout-group': 'ingester' }) +
+    statefulSet.mixin.spec.updateStrategy.withType('OnDelete') +
+    statefulSet.mixin.spec.template.spec.withTerminationGracePeriodSeconds(1200) +
+    statefulSet.mixin.spec.withReplicas(std.ceil($._config.multi_zone_ingester_replicas / 3)) +
+    (if !std.isObject($._config.node_selector) then {} else statefulSet.mixin.spec.template.spec.withNodeSelectorMixin($._config.node_selector)) +
+    if $._config.ingester_allow_multiple_replicas_on_same_node then {} else {
+      spec+:
+        // Allow to schedule 2+ ingesters in the same zone on the same node, but do not schedule 2+ ingesters in
+        // different zones on the same node. In case of 1 node failure in the Kubernetes cluster, only ingesters
+        // in 1 zone will be affected.
+        podAntiAffinity.withRequiredDuringSchedulingIgnoredDuringExecution([
+          podAntiAffinity.requiredDuringSchedulingIgnoredDuringExecutionType.new() +
+          podAntiAffinity.requiredDuringSchedulingIgnoredDuringExecutionType.mixin.labelSelector.withMatchExpressions([
+            { key: 'rollout-group', operator: 'In', values: ['ingester'] },
+            { key: 'name', operator: 'NotIn', values: [name] },
+          ]) +
+          podAntiAffinity.requiredDuringSchedulingIgnoredDuringExecutionType.withTopologyKey('kubernetes.io/hostname'),
+        ]).spec,
+    },
+
+  // Creates a headless service for the per-zone ingesters StatefulSet. We don't use it
+  // but we need to create it anyway because it's responsible for the network identity of
+  // the StatefulSet pods. For more information, see:
+  // https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#statefulset-v1-apps
+  newIngesterZoneService(sts)::
+    $.util.serviceFor(sts, $._config.service_ignored_labels) +
+    service.mixin.spec.withClusterIp('None'),  // Headless.
+
+  ingester_zone_a_container:: if !$._config.multi_zone_ingester_enabled then null else
+    self.newIngesterZoneContainer('a', $.ingester_zone_a_args),
+
+  ingester_zone_a_statefulset: if !$._config.multi_zone_ingester_enabled then null else
```

Comment:
in order to get things to work properly we had to use `{}` instead of `null` for each of these statefulset creations